// @roots/bud-extensions/lib/extensions/cdn/index.js
import { __decorate, __metadata } from "tslib";
import { Extension } from '@roots/bud-framework/extension';
import { bind, disabled, expose, label, options, } from '@roots/bud-framework/extension/decorators';
import isFunction from '@roots/bud-support/lodash/isFunction';
import isString from '@roots/bud-support/lodash/isString';
import isUndefined from '@roots/bud-support/lodash/isUndefined';
/**
 * `@roots/bud-extensions/cdn
 *
 * @remarks
 * Include remote modules in compilation
 *
 * @public
 * @decorator `@label`
 * @decorator `@expose`
 * @decorator `@options`
 * @decorator `@disabled`
 */
let Cdn = class Cdn extends Extension {
    constructor() {
        super(...arguments);
        /**
         * CDN key to URL mapping
         *
         * @public
         */
        this.sources = new Map([
            [`gist`, `https://gist.githubusercontent.com/`],
            [`github`, `https://raw.githubusercontent.com/`],
            [`unpkg`, `https://unpkg.com/`],
            [`skypack`, `https://cdn.skypack.dev/`],
        ]);
        /**
         * Whether to cache modules locally
         *
         * @public
         */
        this.cacheEnabled = true;
    }
    /**
     * Disable local caching of modules
     *
     * @public
     */
    disableCache() {
        this.cacheEnabled = false;
        return this;
    }
    /**
     * Register CDN
     *
     * @public
     */
    registerSource(name, url) {
        this.sources.set(name, url);
        return this;
    }
    /**
     * Allowed URIs getter/setter
     *
     * @public
     */
    get allowedUris() {
        return Array.from(new Set([
            ...this.getOption(`allowedUris`),
            ...(this.sources.values() ?? []),
        ])).filter(v => typeof v === `string` || v instanceof RegExp || isFunction(v));
    }
    set allowedUris(value) {
        this.setOption(`allowedUris`, value);
    }
    /**
     * Cache location getter/setter
     *
     * @public
     */
    get cacheLocation() {
        return this.app.maybeCall(this.getOption(`cacheLocation`));
    }
    set cacheLocation(value) {
        this.setOption(`cacheLocation`, value);
    }
    /**
     * Frozen getter/setter
     *
     * @public
     */
    get frozen() {
        return this.app.maybeCall(this.getOption(`frozen`));
    }
    set frozen(value) {
        this.setOption(`frozen`, value);
    }
    /**
     * Lockfile location getter/setter
     *
     * @public
     */
    get lockfileLocation() {
        return this.getOption(`lockfileLocation`);
    }
    set lockfileLocation(value) {
        this.setOption(`lockfileLocation`, value);
    }
    /**
     * Proxy getter/setter
     *
     * @public
     */
    get proxy() {
        return this.app.maybeCall(this.getOption(`proxy`));
    }
    set proxy(value) {
        this.setOption(`proxy`, value);
    }
    /**
     * Upgrade location getter/setter
     *
     * @public
     */
    get upgrade() {
        return this.app.maybeCall(this.getOption(`upgrade`));
    }
    set upgrade(value) {
        this.setOption(`upgrade`, value);
    }
    /**
     * Set allowed URLs
     *
     * @public
     * @decorator `@bind`
     */
    setAllowedUris(value) {
        this.allowedUris = value;
        return this;
    }
    /**
     * Set cache location
     *
     * @public
     * @decorator `@bind`
     */
    setCacheLocation(value) {
        this.cacheLocation = value;
        return this;
    }
    /**
     * Freeze?
     *
     * @public
     * @decorator `@bind`
     */
    freeze(value) {
        this.frozen = !isUndefined(value) ? value : true;
        return this;
    }
    /**
     * Set lockfile location
     *
     * @public
     * @decorator `@bind`
     */
    setLockfileLocation(value) {
        this.lockfileLocation = value;
        return this;
    }
    /**
     * Set proxy location
     *
     * @public
     * @decorator `@bind`
     */
    setProxy(value) {
        this.proxy = value;
        return this;
    }
    /**
     * Set upgrade
     *
     * @public
     * @decorator `@bind`
     */
    setUpgrade(value) {
        this.upgrade = value;
        return this;
    }
    async beforeBuild() {
        for (const cdnKey of this.sources.keys()) {
            this.logger.log(`registering`, cdnKey);
            this.app.context.manifest?.bud?.[cdnKey] && this.enable();
        }
    }
    /**
     * `buildBefore` callback
     *
     * @public
     * @decorator `@bind`
     */
    async buildBefore(bud) {
        const manifest = bud.context.manifest.bud;
        bud.hooks.on(`build.experiments`, experiments => ({
            ...(experiments ?? {}),
            buildHttp: {
                allowedUris: this.allowedUris.length > 0 ? this.allowedUris : undefined,
                cacheLocation: this.cacheEnabled ? this.cacheLocation : false,
                frozen: this.frozen,
                lockfileLocation: this.lockfileLocation,
                proxy: isString(this.proxy) ? this.proxy : undefined,
                upgrade: this.upgrade,
            },
        }));
        for (const source of this.sources.entries()) {
            const cdn = {
                ident: source[0],
                url: source[1],
                schema: `${source[0]}:`,
            };
            await bud.extensions.add({
                label: `bud-cdn-${cdn.ident}`,
                make: async () => {
                    const { NormalModuleReplacementPlugin } = await import(`webpack`).then(m => m.default);
                    return new NormalModuleReplacementPlugin(new RegExp(`^${cdn.schema}`), result => {
                        result.request = result.request.replace(cdn.schema, cdn.url);
                    });
                },
            });
            const imports = manifest?.imports?.[cdn.ident];
            if (isUndefined(imports))
                return;
            await Promise.all(imports.map(async ([signifier, remotePath]) => {
                await bud.extensions.add({
                    label: `bud-cdn-${cdn.ident}-${remotePath}`,
                    make: async () => {
                        const { NormalModuleReplacementPlugin } = await import(`webpack`).then(m => m.default);
                        return new NormalModuleReplacementPlugin(new RegExp(`^${signifier}`), result => {
                            result.request = result.request.replace(signifier, [cdn.url, remotePath].join(``));
                        });
                    },
                });
            }));
        }
    }
};
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Object)
], Cdn.prototype, "disableCache", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [String, String]),
    __metadata("design:returntype", Object)
], Cdn.prototype, "registerSource", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Object)
], Cdn.prototype, "setAllowedUris", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Object)
], Cdn.prototype, "setCacheLocation", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Boolean]),
    __metadata("design:returntype", Object)
], Cdn.prototype, "freeze", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Object)
], Cdn.prototype, "setLockfileLocation", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Object)
], Cdn.prototype, "setProxy", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Object)
], Cdn.prototype, "setUpgrade", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], Cdn.prototype, "beforeBuild", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Function]),
    __metadata("design:returntype", Promise)
], Cdn.prototype, "buildBefore", null);
Cdn = __decorate([
    label(`@roots/bud-extensions/cdn`),
    expose(`cdn`),
    options({
        allowedUris: [],
        cacheLocation: (app) => app.path(`@storage`, app.label, `modules`),
        frozen: false,
        lockfileLocation: (app) => app.path(`@storage`, app.label, `bud.lock`),
        proxy: ({ env }) => env.isString(`HTTP_PROXY`) && env.get(`HTTP_PROXY`),
        upgrade: true,
    }),
    disabled
], Cdn);
export default Cdn;
